file.go

```diff
@@ -47,14 +48,26 @@ func OpenFile(r io.ReaderAt, size int64, options ...FileOption) (*File, error) {
 		return nil, err
 	}
 
-	if _, err := r.ReadAt(b[:4], 0); err != nil {
+	f.MetadataReaders = c.MetadataReaders
+
+	// check if we should use cached reader for header
+	reader := r
+	if f.MetadataReaders != nil {
+		r = f.MetadataReaders.HeaderReader()
+	}
+	if _, err := reader.ReadAt(b[:4], 0); err != nil {
 		return nil, fmt.Errorf("reading magic header of parquet file: %w", err)
 	}
 	if string(b[:4]) != "PAR1" {
 		return nil, fmt.Errorf("invalid magic header of parquet file: %q", b[:4])
 	}
 
-	if _, err := r.ReadAt(b[:8], size-8); err != nil {
+	// check if we should use cached reader for footer
+	reader = r
+	if f.MetadataReaders != nil {
+		r = f.MetadataReaders.FooterReader()
+	}
+	if _, err := reader.ReadAt(b[:8], size-8); err != nil {
```

Comment:
I wonder whether it would be more meaningful if the "footer reader" expose only the section of the file which contains the footer, in which case the offset wouldn't be `size-8` because that's an absolute position in the original `io.ReaderAt`.

Maybe the `FooterReader() io.ReaderAt` method would be better expressed as `Footer() *io.SectionReader` or `Footer() (r io.ReaderAt, size int64)`.

Reply:
👍  agree with your thoughts on this